cmd/snap-confine/mount-support.c

```diff
@@ -404,6 +404,49 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 			    dst);
 		}
 	}
+	// Since we mountec /etc from the host filesystem the /etc/os-release file
+	// (which may not be present) may be a symbolic link. It is expected that
+	// we see the real /etc/os-release file so let's do our best to show
+	// /etc/os-release as it looks like on the clasic distribution.
+	//
+	// If the symbolic link in /etc is pointing to /usr/lib/os-release then
+	// bind mount /usr/lib/os-release over itself in the core snap. This way
+	// the symlink (which we cannot over-bind mount anything) will work as
+	// expected.
+	//
+	// https://bugs.launchpad.net/canonical-livepatch-client/+bug/1667470
+	if (config->on_classic) {
+		const char *etc_os_release = "/etc/os-release";
+		const char *expected_symlink = "../usr/lib/os-release";
+		const char *usr_lib_os_release = "/usr/lib/os-release";
+		char link_target[PATH_MAX];
+		// Try to read link target from /etc/os-release and special case EINVAL
+		// instead of trying to lstat it first.
+		if (readlink(etc_os_release, link_target, sizeof link_target) <
+		    0) {
+			if (errno != EINVAL) {
+				die("cannot read target of symbolic link from %s", etc_os_release);
+			}
+			goto skip_etc_os_release;
+		}
+		// Check if the link points to ../usr/lib/os-release (or just
+		// /usr/lib/os-release).  If not then we're not ready for that case and
+		// we can just bail out.
+		if (strcmp(link_target, expected_symlink) != 0) {
+			goto skip_etc_os_release;
+		}
```

Comment:
On second though I think realpath will make things harder because we only care about the first level of symbolic link. For example on Fedora the file `/etc/os-release` is a symbolic link to `/usr/lib/os-release` which itself is a symbolic link to something in `/usr/lib/os-release.d/...`.

Our goal is to figure out if we should bind mount, inside the core snap, the immediate target of the link. We don't care where `/etc/os-release` really points to, just that we bind mount the immediate target so that once inside the core snap it will resolve correctly. Since mount follows symbolic links we can just bind mount `/etc/os-release` to `$(readlink /etc/os-release)` and have the desired effect.

I'll rework the code but I don't think realpath buys us anything so I won't use it.